### PR TITLE
feat(nodes): add the Scan Cropper node

### LINKED ISSUE #2014: Scan Cropper node: split a scan holding several photographs into one image per photo
### Problem Statement

A scanned album page holds several photographs in one image, and everything downstream
treats it as one picture: one thumbnail, one embedding, one set of tags, one file in the
store. Digitising a family archive therefore stalls at the point where a person opens each
scan in an editor and cuts the photographs out by hand. One album measured here: 33 pages,
98 photographs.

No node splits an image into its parts. The image nodes we have transform a picture
(`thumbnail`, `background_removal`) or read it (`ocr`, `caption`, `detect`); none turns one
input image into several output images.

Desktop tools solve the cropping but not the pipeline. They need a person driving a GUI,
they write to a local disk, and they leave nothing auditable behind — after a batch there is
no record of what was found, where, or what was rejected.

### Proposed Solution

A filter node on the `image` lane that emits one JPEG per photograph found, in page reading
order, plus a JSON detection record on the `text` lane.

What it decides for itself:

- **Background is inferred from the scan**, so there is no white / dark / album mode to pick.
  A pixel counts as background only when all three hold: it is plain (no local texture), it
  is painted in one of the scan's dominant colours, and it belongs to a region reaching the
  edge of the scan. Each condition covers a gap the others leave.
- **Photos mounted edge to edge are separated** by cutting a region along any straight border
  that spans it, recursively. Standard print ratios arbitrate whether such a cut is believed
  — as a veto, never as a requirement, so a panorama or a trimmed print is still accepted as
  the non-standard shape it is.
- **Detection runs downscaled, crops are cut at full resolution.** That is what makes a
  143 MP flatbed scan tractable at all.

What it guarantees:

- Crops are named `<scan>.crop<N>.jpg` in reading order, numbered densely — a photo detected
  but not cropped leaves no gap, because a gap reads as lost data. The counter runs per
  object, so an upstream fan-out still yields unique names.
- A scan that yields no photographs is **forwarded unchanged**, never dropped.
- `decoded` in the record separates "this is not an image I can read" from "I read it and
  found nothing" — both otherwise report zero regions, and telling them apart is the
  difference between an auditable run and a pile of silent zeros.

Configuration is by **scan type** — Album page, Loose prints, Textured page, Maximum detail —
not by raw numbers. Each preset carries its own advanced panel pre-filled with that preset's
values, so tuning starts from the nearest case and adjusts one thing.

The detection algorithm, the seam search and the JPEG quality matching were developed by the
author in personal forks of [`autocrop`](https://github.com/z80z80z80/autocrop) and
[`ScanCropper`](https://github.com/murniox/ScanCropper), and are contributed here by their
author under this repository's MIT licence. Nothing authored by those projects' other
contributors was carried over.

### Alternatives Considered

- **A white / dark / album background setting.** The setting is a proxy for something the
  node can measure directly, and any album has pages that break the mode chosen for the batch.
- **Requiring standard print ratios.** Would discard panoramas and trimmed prints. Ratios
  arbitrate ambiguous cuts instead, and never reject a photo on their own.
- **Extending an existing image node.** None of them fans out, and the fan-out is the feature.
- **Leaving it to a desktop tool outside the pipeline.** Loses provenance and the audit
  record, and cannot be scheduled.

### Affected Modules

- [x] nodes (pipeline)
- [x] ai

### Depends on

Needs #1941. Until the shared node base settles media streams, a consumer keeps only the last
crop of each scan, so the node's output is silently discarded downstream.

### PR DESCRIPTION
**Pull request by asclearuc**
## Summary

- A scanned album page holds several photographs, and every node downstream treats it as one picture: one thumbnail, one embedding, one file in the store. Splitting them was a manual step outside the pipeline — 33 pages of one album carry 98 photographs.
- Adds a filter node that emits one JPEG per photograph in page reading order, with a JSON detection record on the `text` lane saying what was found, where, and what it was called. It infers the background from the scan instead of asking which kind of page you have, separates photos mounted edge to edge, and detects on a downscaled copy so a 143 MP flatbed scan is tractable while the crops are still cut at full resolution.
- Moves `source_quality` / `matched_quality` into `ai.common.utils`. They recover the quality a JPEG was written at and pick one to re-encode with, need only the standard library, and the orientation node coming next uses the first of them.

## Type

feature

## Ordering

Nothing blocks it. The shared media-lane normalization this node depends on — a producer of
several streams per object needs it, or a consumer keeps only the last crop — merged in #1962
and is on `develop`. The branch is rebased on top of it, so the numbers below are from a run
against that base rather than against a `develop` that could not carry the fan-out.

## Testing

- [x] Tests added or updated
- [x] Tested locally
- [x] `./builder test` passes

68 tests. Fan-out and naming, reading order, the audit record, the no-photos path and preset
resolution; the detector core itself — two prints mounted edge to edge come apart, and the
same scan with `maxDepth 0` stays one region, so the seam logic is pinned in both directions;
and the config layer, where a value that is not a finite number, or is a bool, falls back
rather than travelling on as a setting nobody asked for.

`./builder test` on 92f4ff408, rebased onto `develop` with #1962 in it. The tip has moved
one commit since, touching only a test file:

```
nodes              3427 passed,  221 skipped   (4:37)
ai                 1997 passed,  122 skipped   (5:12)
client-python       121 passed,    6 skipped   (9:09)
client-typescript   215 passed,    8 skipped
client-mcp           75 passed
rocketlib            27 passed
server, shared, shell, docs   passed
```

End to end over a 33-page album: 98 crops, matching a manual count of the photographs on
those pages. That manual run, and the declarative `count == 3` case, are what cover the
detector's *judgement*; the unit tests cover its plumbing.

## Checklist

- [x] Commit messages follow [conventional commits](https://www.conventionalcommits.org/)
- [x] No secrets or credentials included
- [ ] Wiki updated (if applicable)
- [x] Breaking changes documented (if applicable)

## Provenance

The detection algorithm, the seam search and the JPEG quality tables were developed by the
author in personal forks of [`autocrop`](https://github.com/z80z80z80/autocrop) and
[`ScanCropper`](https://github.com/murniox/ScanCropper), and are contributed here by their
author under this repository's MIT licence. Nothing authored by those projects' other
contributors was carried over. Commits are cited in the node's README.

## Limitations, stated rather than hidden

- Like every other buffering node here, it has no input size cap. Decoding a 143 MP scan
  costs roughly 430 MB, plus about 105 MB per crop while it is encoded.
- Crops are re-encoded with OpenCV, which writes no EXIF and no ICC profile. They are new
  images, not edited originals.
- Edge accuracy is bounded by `detectSize`: at the default, one pixel of detection error is
  about five pixels on a 14000 px-tall scan.

## Linked Issue

Closes #2014

## New Features
- Added scan cropping to detect, split, order, and export photos from scanned images.
- Added configurable scan-type presets, detection settings, output quality, and JSON region metadata.
- Added handling for unreadable scans and scans without detected photos.
- Added JPEG quality detection and matching utilities.

## Documentation
- Added comprehensive scan-cropping configuration, usage, troubleshooting, and limitations guidance.

## Tests
- Added coverage for cropping, ordering, presets, metadata, quality handling, and edge cases.

### CODE REVIEW

File: nodes/src/nodes/scan_cropper/detect.py [hunk 1]
```diff
@@ -0,0 +1,707 @@
...
+
+    img16 = img.astype(np.int16)
+    bg = np.zeros((h, w), dtype=bool)
+    for c in peaks:
+        same = plain & ~wall & (np.abs(img16 - unpack_colour(c)).max(axis=2) < COLOUR_TOL)
+        _, labels = cv2.connectedComponents(same.astype(np.uint8), 8)
+        at_border = set(labels[0, :]) | set(labels[-1, :]) | set(labels[:, 0]) | set(labels[:, -1])
+        at_border.discard(0)
+        if at_border:
+            bg |= np.isin(labels, list(at_border))
```
[the remark this answers is not in the document]
Comment (asclearuc) [reply nodes/src/nodes/scan_cropper/detect.py:229]: @coderabbit 
I checked this on the engine before changing it. OpenCV does not reject the call:

```
opencv 4.13.0
POSITIONAL 8 ACCEPTED; labels shape (4, 4) n 2
keyword form n 2
```

The second positional parameter is the output `labels` array, and OpenCV accepts a scalar there. Connectivity then falls back to its default, which is 8 — the same value the code meant to pass. So the result was already correct, and nothing crashed. The severity here is not critical.

I made the change anyway, in 46d5a9786. The call reads as if it sets connectivity and it does not, which is a trap for the next reader. But it is a readability fix, not a bug fix.

File: nodes/src/nodes/scan_cropper/services.json [hunk 2]
```diff
@@ -0,0 +1,514 @@
...
+	//
+	// Optional:
+	//		Description to of this driver
+	//
+	"description": ["Splits a scan that holds several photos - a photo album page, or a handful of prints laid on the", "platen - into one image per photo. It works out what the background is from the scan itself, so there", "is no white/dark/album mode to choose, straightens each photo, and cuts apart prints mounted edge to", "edge. Detection runs on a downscaled copy and the crops are taken at full resolution."],
+	//
+	// Optional:
+	//	  The icon is the icon to display in the UI for this node
+	//
+	"icon": "scan_cropper.svg",
```
[the remark this answers is not in the document]
Comment (asclearuc) [reply nodes/src/nodes/scan_cropper/services.json:57]: @coderabbit 
The fixture is in the PR now. `testdata/images/scan_album_3up.png` was added in 13d02454, about ten minutes after this comment was written.

The path in `services.json` is relative to `testdata/`, the same as `images/einstein.jpg` in `background_removal` and `caption`.

File: packages/ai/src/ai/common/utils/image_utils.py [hunk 3]
```diff
@@ -156,3 +162,200 @@ def scale_point(point: Dict[str, float], fx: float, fy: float) -> None:
...
+    i = 2
+    while i < len(data) - 3:
+        if data[i] != 0xFF:
+            i += 1
+            continue
+        marker = data[i + 1]
+        if marker == 0xD8 or 0xD0 <= marker <= 0xD7:  # no payload
+            i += 2
+            continue
+        if marker in (0xDA, 0xD9):  # pixel data starts
```
[the remark this answers is not in the document]
Comment (asclearuc) [reply packages/ai/src/ai/common/utils/image_utils.py:320]: @coderabbit 
recheck

File: nodes/test/scan_cropper/test_fanout.py [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+
+    def test_undecodable_forwards_the_original(self):
+        """Same for bytes we could not decode: forward them on rather than swallowing them."""
+        inst = _make(None)
+
+        _send_stream(inst, _descriptor_payload(), body=b'original-scan-bytes')
+
+        triplets = _triplets(inst.instance)
+        assert len(triplets) == 1
+        assert triplets[0][1][2] == b'original-scan-bytes'
```
[the remark this answers is not in the document]
Comment (asclearuc) [reply nodes/test/scan_cropper/test_fanout.py:294]: @coderabbit 
recheck

File: nodes/src/nodes/scan_cropper/process.py [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+            return None
+
+        height, width = img.shape[:2]
+        page_area = float(height) * float(width)
+
+        rects = reading_order(find_photos(img, params))
+        regions = [_region(r, page_area) for r in rects]
+
+        if not want_images:
+            return [], regions
```
[the remark this answers is not in the document]
Comment (asclearuc) [reply nodes/src/nodes/scan_cropper/process.py:238]: @coderabbit 
recheck

File: nodes/src/nodes/scan_cropper/process.py [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+        str | int: ``'auto'``, or a quality clamped to 1..100.
+    """
+    text = str(value).strip().lower()
+    if text == 'auto':
+        return 'auto'
+    try:
+        return max(1, min(100, int(float(text))))
+    except (TypeError, ValueError):
+        warning(f'scan_cropper: config quality={value!r} is neither "auto" nor a number; using auto')
+        return 'auto'
```
[the remark this answers is not in the document]
Comment (asclearuc) [reply nodes/src/nodes/scan_cropper/process.py:128]: @coderabbit 
recheck

File: packages/ai/src/ai/common/utils/image_utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (asclearuc) [reply packages/ai/src/ai/common/utils/image_utils.py:320]: @coderabbit 
recheck

[the remark this answers is not in the document]
Comment (asclearuc) [reply packages/ai/src/ai/common/utils/image_utils.py:320]: @coderabbit 
recheck all your comments and resolve solved ones

File: nodes/src/nodes/scan_cropper/services.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (asclearuc) [reply nodes/src/nodes/scan_cropper/services.json:57]: @coderabbit - is all good ?

[the remark this answers is not in the document]
Comment (asclearuc) [reply nodes/src/nodes/scan_cropper/services.json:57]: @CodeRabbit - is all good ?

File: nodes/src/nodes/scan_cropper/detect.py [hunk 7]
```diff
@@ -0,0 +1,708 @@
...
+            area_k = k[1][0] * k[1][1]
+            if overlap / max(min(area_r, area_k), 1) > overlap_frac:
+                duplicate = True
+                break
+        if not duplicate:
+            kept.append(r)
+    return kept
+
+
+def find_photos(img, params: DetectParams):
```
Comment (dsapandora) [inline nodes/src/nodes/scan_cropper/detect.py:656]: The plumbing side is well covered (fanout, ordering, presets, the JPEG-quality tables against real fixtures). The gap that's left sits right here: the detector core — `foreground_mask`, `best_seam`/`decompose`, `snap_to_border`, `dedupe`, `_accept_split` — has no automated coverage, because the `nodes/test` interpreter carries no cv2. The only thing exercising any of it is the single declarative `count == 3` case, and `./builder test` is still unchecked in the description, so a regression in the seam logic would pass CI silently.

**Review by dsapandora — DISMISSED** [review]
Approving — The listener gating (only decode/encode when something downstream is actually listening), the forward-unchanged fallback on every failure path, and the preset↔profile drift guard in `test_presets.py` are exactly the right calls, and the audit record telling "couldn't read it" apart from "read it, found nothing" is a nice touch. One non-blocking follow-up left inline on `detect.py`.

File: nodes/src/nodes/scan_cropper/detect.py (hunk 7 above)
Comment (asclearuc) [reply nodes/src/nodes/scan_cropper/detect.py:656]: Fair, and it is now covered — `nodes/test/scan_cropper/test_detect.py`, 11 tests. The suite for this node goes 39 → 50.

One correction on the reason, because it changes what the fix looks like: the interpreter does carry cv2.

```
ai.common.opencv OK in 3.3s, cv2 4.13.0
detect.py imported OK
find_photos -> 3 regions
```

No test imported it before, so it looked absent — `ordering.py` is pure geometry and the other files stub `split_scan` out. Nothing was blocking these tests except that I had not written them.

They are synthetic rather than fixture-backed. A scan on disk pins one photograph of one album; a generated page states the property. The pair that matters for the seam logic:

- two prints mounted edge to edge, no paper between them, only a tonal step -> **2 regions**
- the same scan with `maxDepth 0` -> **1 region**

Same image, same code; the only thing deciding the outcome is the setting the README calls the main speed lever. Also covered: a hairline of paper between prints, a blank page, specks under `minArea`, an over-long strip rejected by `maxAspect`, `dedupe` on an overlapping pair and on a disjoint one, and `snap_to_border` declining below its size floor.

One that is not in your list but sat in the same blind spot: detection runs downscaled and the crops are cut at full size, so a `detect_size=400` run is compared against one that did not downscale. Losing the scale-up would take a third-size corner out of every photograph.

Worth saying plainly: this tests the plumbing of the detector, not its judgement. Whether it finds the right rectangles on a real 143 MP album page is still the declarative case and the manual run (33 pages, 98 photographs).

On `./builder test` — running it now, and I will report the result on the PR rather than just ticking the box.

Comment (asclearuc) [reply nodes/src/nodes/scan_cropper/detect.py:656]: `./builder test` — green on 827de7da5, the commit this PR is at. Nothing was touched during the run.

```
nodes              3370 passed,  221 skipped   (4:21)
ai                 1982 passed,  122 skipped   (5:23)
client-python       121 passed,    6 skipped   (9:45)
client-typescript   215 passed,    8 skipped
client-mcp           75 passed
server, shared, shell, docs   passed
```

The 50 tests for this node are inside the `nodes` figure — 39 before the detector file, 50 after.

The checkbox in the description is ticked now, with these numbers written next to it rather than a bare tick.

File: nodes/src/nodes/scan_cropper/process.py [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+        The clamped value, or ``default`` when the config holds something non-numeric.
+    """
+    raw = config.get(key, default)
+    try:
+        value = cast(raw)
+    except (TypeError, ValueError):
+        warning(f'scan_cropper: config {key}={raw!r} is not a number; using {default}')
+        return default
+    low, high = _RANGES[key]
+    return cast(max(low, min(high, value)))
```
[the remark this answers is not in the document]
Comment (asclearuc) [reply nodes/src/nodes/scan_cropper/process.py:102]: Right, and thank you — I widened `resolve_quality` for exactly this and left `_number` behind. Fixed in 148ed77a3.

While there I found the quieter half of it. `OverflowError` only fires on the integer fields, because it takes `int(float('inf'))` to raise. On a float field `float('inf')` succeeds, and `nan` succeeds everywhere — and neither is stopped by the clamp, because every comparison against `nan` is false, so `max(low, min(high, nan))` hands it straight back. It then travels into `DetectParams`, and each threshold reading it becomes a no-op. That one does not fail the node; it makes detection go strange.

So the cast now catches `OverflowError` as well, and anything that survives the cast is checked with `math.isfinite` before it is clamped.

The config layer had no tests at all, so there are 17 now: `_number` over good, missing, out-of-range, non-numeric and non-finite values; `_params_from_config` showing that one poisoned field costs only itself; and `resolve_quality` across text, JSON numbers, clamping and both non-finite forms.

`nodes:test -k scan_cropper` is 67 passed.

[the remark this answers is not in the document]
Comment (asclearuc) [reply nodes/src/nodes/scan_cropper/process.py:102]: Correction on the reference above: that commit was amended while trimming test docstrings, so the fix is now `ac57775a4`, not `148ed77a3`. Same content.

File: nodes/src/nodes/scan_cropper/process.py [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+        value = cast(raw)
+    except (TypeError, ValueError, OverflowError):
+        # OverflowError is not a ValueError: `int(float('inf'))` raises it.
+        warning(f'scan_cropper: config {key}={raw!r} is not a number; using {default}')
+        return default
+    if not math.isfinite(value):
+        # Clamping cannot rescue these: every comparison against nan is false, so it would
+        # travel on into DetectParams and turn each threshold that reads it into a no-op.
+        warning(f'scan_cropper: config {key}={raw!r} is not a finite number; using {default}')
+        return default
```
[the remark this answers is not in the document]
Comment (asclearuc) [reply nodes/src/nodes/scan_cropper/process.py:100]: Agreed, and fixed in 6052d9616.

It is the same shape as the `nan` case rather than a lesser one: `int(True)` is 1, so `"maxDepth": true` passed the cast, the finite check and the clamp, and landed in `DetectParams` as 1. The node would have cut a joined blob apart once instead of four times and said nothing — a different setting, not a crash.

The check has to sit before the cast, because afterwards there is nothing left to tell `True` from `1`.

Worth noting this rule already exists a layer down: `_avi_declared_size` in rocketlib rejects a bool the same way when it reads a declared stream size. I did not carry it across to this node.

Covered on both an integer and a float field. `nodes:test -k scan_cropper` is 68 passed.

File: nodes/test/scan_cropper/test_config.py [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+    def test_every_tunable_survives_a_config_of_rubbish(self):
+        rubbish = dict.fromkeys(
+            ('detectSize', 'texture', 'minArea', 'maxArea', 'maxAspect', 'minRelative', 'maxDepth', 'skew'),
+            'not a number at all',
+        )
+
+        params = _params_from_config(rubbish)
+
+        assert params.skew > 0, 'the seam search divides by this'
+        assert all(math.isfinite(v) for v in (params.texture, params.min_area, params.max_area))
```
[the remark this answers is not in the document]
Comment (asclearuc) [reply nodes/test/scan_cropper/test_config.py]: Right — eight names where `_params_from_config` fills nine, so `ratioTolerance` never saw a bad value and the test claimed more than it checked. Fixed in 14d043850.

I did not add the missing name, because that leaves the same trap for the next tunable someone adds. Both ends are derived now: the rubbish config is built from `_RANGES`, and the assertion walks `dataclasses.fields(params)`. A tenth parameter is covered without touching this test.

`nodes:test -k scan_cropper` is 68 passed.

File: nodes/src/nodes/scan_cropper/detect.py (hunk 7 above)
Comment (dsapandora) [reply nodes/src/nodes/scan_cropper/detect.py:656]: You're right on both counts — cv2 is in that interpreter, the tests just hadn't been written, and the downscale/scale-back check was the real blind spot. test_detect.py covers exactly what mattered here (the edge-to-edge split vs maxDepth=0 pair especially), and `./builder test` green with the numbers written down is what I was after. Nothing else from me — approval stands.

**Review by dsapandora — APPROVED** [review]
Follow-up is resolved — detector core is now covered by test_detect.py and `./builder test` is green on this head. Approving.
